File: go/iter.go
package jrif

import "context"

// ArrayIter iterates over the items of an array cursor in source order. Built
// by [Cursor.Iter], which resolves the array length once (possibly fetching
// and parsing) before handing back the iterator.
//
// When the array's chunk index fully describes its items (the common case),
// Next walks chunks sequentially in O(1) amortized per call. When chunks are
// absent or the cursor is wrapping a chunk fragment, Next falls back to
// [Cursor.Index] per ordinal — correct but with worse asymptotics.
type ArrayIter struct {
	root *Cursor
	len  uint64
	next uint64
	walk *arrayWalker // nil → slow path (per-ordinal Cursor.Index)
}

// Len returns the total number of items in the array.
func (it *ArrayIter) Len() uint64 { return it.len }

// Next yields the cursor for the next item. ok is false when iteration is
// complete.
func (it *ArrayIter) Next() (c *Cursor, ok bool) {
	if it.next >= it.len {
		return nil, false
	}
	if it.walk != nil {
		hit, ord, ok := it.walk.next()
		if ok {
			it.next++
			return it.root.applyArrayHit(hit, ord), true
		}
		// Walker exhausted unexpectedly (len disagreed with chunks). Disable
		// the walker and fall through to the slow path.
		it.walk = nil
	}
	c = it.root.Index(it.next)
	it.next++
	return c, true
}

// Reset rewinds the iterator to position 0.
func (it *ArrayIter) Reset() {
	it.next = 0
	if it.walk != nil {
		it.walk = newArrayWalker(&it.root.frame)
	}
}

// ObjectIter iterates over the (field name, child cursor) pairs of an object
// cursor in source order. Built by [Cursor.Entries].
//
// When the object's chunk index fully describes its fields (the common case),
// Next resolves each child without rescanning chunks. When the chunk index
// can't describe the object (deferred / wrapped path), Entries parses the
// bytes to recover key order and Next falls back to [Cursor.Get] per key.
type ObjectIter struct {
	root    *Cursor
	entries []objectEntry // pre-walked when chunk-described; otherwise name-only
	next    int
}

// Len returns the total number of fields.
func (it *ObjectIter) Len() int { return len(it.entries) }

// Next yields the name and child cursor for the next field. ok is false when
// iteration is complete.
func (it *ObjectIter) Next() (name string, c *Cursor, ok bool) {
	if it.next >= len(it.entries) {
		return "", nil, false
	}
	e := it.entries[it.next]
	it.next++
	if e.hit.kind == objectHitNone {
		return e.name, it.root.Get(e.name), true
	}
	return e.name, it.root.applyObjectHit(e.hit, e.name), true
}

// Reset rewinds the iterator to position 0.
func (it *ObjectIter) Reset() { it.next = 0 }

// Iter resolves the array length, then returns an iterator that yields one
// child Cursor per element. ctx governs any I/O performed to resolve the
// length (fast path: zero fetches).
func (c *Cursor) Iter(ctx context.Context) (*ArrayIter, error) {
	n, err := c.Len(ctx)
	if err != nil {
		return nil, err
	}
	return &ArrayIter{root: c, len: n, walk: newArrayWalker(&c.frame)}, nil
}

// Entries resolves the object's field names in source order, then returns an
// iterator that yields (name, child cursor) pairs. ctx governs the fetch +
// parse required to recover key order (object chunks may cover only a subset
// of fields per spec, so we can't enumerate keys from the chunk index alone).
//
// For each key that is described by an object chunk, the iterator descends
// via that chunk without rescanning. Uncovered fields fall back to
// [Cursor.Get], which itself defers to a parse on first I/O.
func (c *Cursor) Entries(ctx context.Context) (*ObjectIter, error) {
	keys, err := c.objectKeys(ctx)
	if err != nil {
		return nil, err
	}
	var hitsByName map[string]objectHit
	if len(c.pending) == 0 && c.frame.kind == frameObject {
		hitsByName = chunkHitsByName(&c.frame)
	}
	entries := make([]objectEntry, len(keys))
	for i, k := range keys {
		entries[i].name = k
		if hitsByName != nil {
			entries[i].hit = hitsByName[k]
		}
	}
	return &ObjectIter{root: c, entries: entries}, nil
}

File: rust/src/reader/iter.rs
//! [`ArrayIter`] and [`ObjectIter`] — sync iteration over container cursors.
//!
//! `iter().await?` (arrays) and `entries().await?` (objects) each resolve
//! their container shape once, possibly fetching+parsing, then yield child
//! cursors infallibly via std `Iterator`.

use crate::nav::{ArrayWalker, Frame};
use crate::reader::Cursor;
use crate::reader::source::Source;

/// Iterator returned by [`Cursor::iter`](crate::Cursor::iter).
///
/// When the array is described by chunks in the index, items are emitted via
/// an [`ArrayWalker`] for O(1) amortized work per step (no per-item chunk
/// rescans). When the iterator's root cursor has pending segments or isn't
/// chunk-described (e.g. an inline array), iteration falls back to per-ordinal
/// `Cursor::index`, which is correct but asymptotically worse.
pub struct ArrayIter<'a, F> {
    pub(crate) root: Cursor<'a, F>,
    pub(crate) len: u64,
    pub(crate) next: u64,
    /// Walker over the root's array chunks, when available. We hold it as
    /// `Option<...>` so the slow-path (no chunk frame, e.g. inline arrays) is
    /// also expressible.
    pub(crate) walker: Option<ArrayWalker<'a>>,
}

impl<'a, F: Source> Iterator for ArrayIter<'a, F> {
    type Item = Cursor<'a, F>;

    fn next(&mut self) -> Option<Self::Item> {
        if self.next >= self.len {
            return None;
        }
        if let Some(w) = self.walker.as_mut()
            && let Some((hit, ord)) = w.next()
        {
            self.next += 1;
            return Some(self.root.clone().apply_array_hit(hit, ord));
        }
        let ord = self.next;
        self.next += 1;
        Some(self.root.clone().index(ord))
    }

    fn size_hint(&self) -> (usize, Option<usize>) {
        let remaining = usize::try_from(self.len - self.next).expect("array length fits in usize");
        (remaining, Some(remaining))
    }
}

impl<F: Source> ExactSizeIterator for ArrayIter<'_, F> {}
impl<F: Source> std::iter::FusedIterator for ArrayIter<'_, F> {}

impl<'a, F> ArrayIter<'a, F> {
    /// Build a new iterator. Used by [`Cursor::iter`]; selects the walker
    /// fast path when the root cursor exposes a chunk-described array frame.
    pub(crate) fn new(root: Cursor<'a, F>, len: u64) -> Self {
        let walker = if root.pending.is_empty() {
            match root.frame {
                Frame::Array(chunks) => Some(ArrayWalker::new(chunks)),
                _ => None,
            }
        } else {
            None
        };
        Self {
            root,
            len,
            next: 0,
            walker,
        }
    }
}

/// Iterator returned by [`Cursor::entries`](crate::Cursor::entries).
/// Yields `(field_name, cursor)` pairs in source order.
pub struct ObjectIter<'a, F> {
    pub(crate) root: Cursor<'a, F>,
    pub(crate) keys: std::vec::IntoIter<Box<str>>,
}

impl<'a, F: Source> Iterator for ObjectIter<'a, F> {
    type Item = (Box<str>, Cursor<'a, F>);

    fn next(&mut self) -> Option<Self::Item> {
        let key = self.keys.next()?;
        let cursor = self.root.clone().get(&key);
        Some((key, cursor))
    }

    fn size_hint(&self) -> (usize, Option<usize>) {
        let r = self.keys.len();
        (r, Some(r))
    }
}

impl<F: Source> ExactSizeIterator for ObjectIter<'_, F> {}
impl<F: Source> std::iter::FusedIterator for ObjectIter<'_, F> {}
